File: src/main/java/fr/lewon/bot/manager/rest/BotController.java
package fr.lewon.bot.manager.rest;

import java.io.IOException;

import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.util.StreamUtils;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import fr.lewon.bot.manager.entities.BotInfosDTO;
import fr.lewon.bot.manager.entities.BotLogsDTO;
import fr.lewon.bot.manager.entities.BotMethodsDTO;
import fr.lewon.bot.manager.entities.BotPropertiesDescriptorsDTO;
import fr.lewon.bot.manager.entities.CallMethodDTO;
import fr.lewon.bot.manager.entities.GameInfosListDTO;
import fr.lewon.bot.manager.entities.UserInfosDTO;
import fr.lewon.bot.manager.service.BotService;
import fr.lewon.bot.manager.util.errors.BotManagerException;

@CrossOrigin
@RestController
@RequestMapping("/bots")
public class BotController {

	@Autowired
	private BotService botService;

	@GetMapping(produces = "application/json", value = "/all")
	public ResponseEntity<GameInfosListDTO> getAllBotInfos() throws BotManagerException {
		GameInfosListDTO infos = botService.getAllBotInfos();
		return new ResponseEntity<>(infos, HttpStatus.OK);
	}

	@GetMapping(produces = "application/json", value = "/{gameId}/bot/{id}")
	public ResponseEntity<BotInfosDTO> getBotInfos(@PathVariable String gameId, @PathVariable Long id) throws BotManagerException {
		BotInfosDTO infos = botService.getBotInfos(gameId, id);
		return new ResponseEntity<>(infos, HttpStatus.OK);
	}
	
	@PostMapping(produces = "application/json", consumes = "application/json", value = "/{gameId}/create")
	public ResponseEntity<Void> createBot(@PathVariable String gameId, @RequestBody UserInfosDTO infos) throws BotManagerException {
		botService.createBot(infos.getLogin(), infos.getPassword(), gameId, infos.getParams());
		return new ResponseEntity<>(HttpStatus.OK);
	}

	@PostMapping(produces = "application/json", value = "/{gameId}/bot/{id}/stop")
	public ResponseEntity<Void> stopBot(@PathVariable String gameId, @PathVariable Long id) throws BotManagerException {
		botService.stopBot(gameId, id);
		return new ResponseEntity<>(HttpStatus.OK);
	}
	
	@PostMapping(produces = "application/json", value = "/{gameId}/bot/{id}/start")
	public ResponseEntity<Void> startBot(@PathVariable String gameId, @PathVariable Long id) throws BotManagerException {
		botService.startBot(gameId, id);
		return new ResponseEntity<>(HttpStatus.OK);
	}

	@PostMapping(produces = "application/json", value = "/{gameId}/bot/{id}/pause")
	public ResponseEntity<Void> pauseBot(@PathVariable String gameId, @PathVariable Long id) throws BotManagerException {
		botService.pauseBot(gameId, id);
		return new ResponseEntity<>(HttpStatus.OK);
	}

	@PostMapping(produces = "application/json", value = "/{gameId}/bot/{id}/trim")
	public ResponseEntity<Void> trimStoppedBot(@PathVariable String gameId, @PathVariable Long id) throws BotManagerException {
		botService.trimStoppedBot(gameId, id);
		return new ResponseEntity<>(HttpStatus.OK);
	}

	@PostMapping(produces = "application/json", value = "/trim")
	public ResponseEntity<Void> trimStoppedBots() throws BotManagerException {
		botService.trimStoppedBots();
		return new ResponseEntity<>(HttpStatus.OK);
	}

	@GetMapping(produces = "application/json", value = "/{gameId}/bot/{id}/logs")
	public ResponseEntity<BotLogsDTO> getLogs(@PathVariable String gameId, @PathVariable Long id) throws BotManagerException {
		BotLogsDTO logs = botService.getBotLogs(gameId, id);
		return new ResponseEntity<>(logs, HttpStatus.OK);
	}

	@GetMapping(produces = "application/json", value = "/{gameId}/bot/{botId}/methods")
	public ResponseEntity<BotMethodsDTO> getMethods(@PathVariable String gameId, @PathVariable Long botId) throws BotManagerException {
		BotMethodsDTO methods = botService.getMethods(gameId, botId);
		return new ResponseEntity<>(methods, HttpStatus.OK);
	}

	@PostMapping(produces = "application/json", consumes = "application/json", value = "/{gameId}/bot/{botId}/methods/call")
	public ResponseEntity<Object> callMethod(@PathVariable String gameId, @PathVariable Long botId, @RequestBody CallMethodDTO callMethodDTO) throws BotManagerException {
		Object result = botService.callMethod(gameId, callMethodDTO.getMethodId(), botId, callMethodDTO.getParams());
		return new ResponseEntity<>(result, HttpStatus.OK);
	}
	
	@GetMapping(produces = "application/json", value = "/{gameId}/properties")
	public ResponseEntity<BotPropertiesDescriptorsDTO> getParams(@PathVariable String gameId) throws BotManagerException {
		BotPropertiesDescriptorsDTO props = botService.getProperties(gameId);
		return new ResponseEntity<>(props, HttpStatus.OK);
	}

	@GetMapping(produces = MediaType.IMAGE_JPEG_VALUE, value = "/{gameId}/icons")
	public void getImage(@PathVariable String gameId, HttpServletResponse response) throws BotManagerException {
			ClassPathResource imgFile = botService.getIcon(gameId);
		response.setContentType(MediaType.IMAGE_JPEG_VALUE);
		try {
			StreamUtils.copy(imgFile.getInputStream(), response.getOutputStream());
		} catch (IOException e) {
			throw new BotManagerException(e.getMessage());
		}
	}
}

File: src/main/kotlin/fr/lewon/bot/manager/rest/BotController.kt
package fr.lewon.bot.manager.rest

import fr.lewon.bot.manager.entities.*
import fr.lewon.bot.manager.service.BotService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.util.StreamUtils
import org.springframework.web.bind.annotation.*
import javax.servlet.http.HttpServletResponse

@CrossOrigin
@RestController
@RequestMapping("/bots")
open class BotController {

    @Autowired
    private lateinit var botService: BotService

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/all"])
    fun getAllBotInfo(): ResponseEntity<GameInfoListDTO> {
        val info = botService.getAllBotInfo()
        return ResponseEntity(info, HttpStatus.OK)
    }

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/bot/{id}"])
    fun getBotInfo(@PathVariable id: Long): ResponseEntity<BotInfoDTO> {
        val info = botService.getBotInfo(id)
        return ResponseEntity(info, HttpStatus.OK)
    }

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/bot/{id}/tasks"])
    fun getBotTasks(@PathVariable id: Long): ResponseEntity<BotTasksDTO> {
        val tasks = botService.getBotTasks(id)
        return ResponseEntity(tasks, HttpStatus.OK)
    }

    @PostMapping(
        produces = [MediaType.APPLICATION_JSON_VALUE],
        consumes = [MediaType.APPLICATION_JSON_VALUE],
        value = ["/{gameId}/create"]
    )
    fun createBot(@PathVariable gameId: Long, @RequestBody info: UserInfoDTO): ResponseEntity<BotInfoDTO> {
        val botInfo = botService.createBot(info.login, info.loginProperties, gameId, info.params)
        return ResponseEntity(botInfo, HttpStatus.OK)
    }

    @PostMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/bot/{id}/transition/{transition}"])
    fun processBotTransition(@PathVariable transition: String, @PathVariable id: Long): ResponseEntity<Nothing> {
        botService.processBotTransition(transition, id)
        return ResponseEntity(HttpStatus.OK)
    }

    @PostMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/bot/{id}/trim"])
    fun trimStoppedBot(@PathVariable id: Long): ResponseEntity<Nothing> {
        botService.trimStoppedBot(id)
        return ResponseEntity(HttpStatus.OK)
    }

    @PostMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/trim"])
    fun trimStoppedBots(): ResponseEntity<Nothing> {
        botService.trimStoppedBots()
        return ResponseEntity(HttpStatus.OK)
    }

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/bot/{id}/logs"])
    fun getLogs(@PathVariable id: Long): List<BotLogDTO> {
        return botService.getBotLogs(id)
    }

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/bot/{botId}/operations"])
    fun getBotOperations(@PathVariable botId: Long): ResponseEntity<BotOperationsDTO> {
        val methods = botService.getBotOperations(botId)
        return ResponseEntity(methods, HttpStatus.OK)
    }

    @PostMapping(
        produces = [MediaType.APPLICATION_JSON_VALUE],
        consumes = [MediaType.APPLICATION_JSON_VALUE],
        value = ["/bot/{botId}/operation/{operationId}/call"]
    )
    fun callBotOperation(
        @PathVariable botId: Long,
        @PathVariable operationId: Long,
        @RequestBody params: Map<String, String?>
    ): ResponseEntity<OperationResultDTO> {
        val result = botService.callBotOperation(operationId, botId, params)
        return ResponseEntity(result, HttpStatus.OK)
    }

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/{gameId}/properties"])
    fun getGameProperties(@PathVariable gameId: Long): ResponseEntity<BotPropertiesDescriptorsDTO> {
        val props = botService.getGameProperties(gameId)
        return ResponseEntity(props, HttpStatus.OK)
    }

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/{gameId}/loginProperties"])
    fun getGameLoginProperties(@PathVariable gameId: Long): ResponseEntity<BotPropertiesDescriptorsDTO> {
        val properties = botService.getLoginProperties(gameId)
        return ResponseEntity(properties, HttpStatus.OK)
    }

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/bot/{botId}/properties"])
    fun getBotProperties(@PathVariable botId: Long): ResponseEntity<Map<String, Any?>> {
        val properties = botService.getBotProperties(botId)
        return ResponseEntity(properties, HttpStatus.OK)
    }

    @PostMapping(produces = [MediaType.APPLICATION_JSON_VALUE], value = ["/bot/{botId}/properties"])
    fun updateBotProperties(
        @PathVariable botId: Long,
        @RequestBody properties: Map<String, String?>
    ): ResponseEntity<Map<String, Any?>> {
        val newProperties = botService.updateBotProperties(botId, properties)
        return ResponseEntity(newProperties, HttpStatus.OK)
    }

    @GetMapping(produces = [MediaType.IMAGE_JPEG_VALUE], value = ["/{gameId}/icons"])
    fun getImage(@PathVariable gameId: Long, response: HttpServletResponse) {
        val imgFile = botService.getIcon(gameId)
        response.contentType = MediaType.IMAGE_JPEG_VALUE
        StreamUtils.copy(imgFile.inputStream, response.outputStream)
    }
}